ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.utils
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Future, Promise}
+
+trait ConcurrencyLimiter {
+  def execute[T](task: () => Future[T]): Future[T]
+}
+
+class NoConcurrencyLimiter extends ConcurrencyLimiter {
+  override def execute[T](task: () => Future[T]): Future[T] = task()
+}
+
+class QueueBasedConcurrencyLimiter(
+    parallelism: Int,
+    executionContext: ExecutionContext,
+) extends ConcurrencyLimiter {
+  assert(parallelism > 0)
+
+  type Task = () => Unit
+  private val waiting = mutable.Queue[Task]()
+  private var running: Int = 0
+
+  override def execute[T](task: () => Future[T]): Future[T] = synchronized {
```

Comment:
enqueueTask with a startTask() as the before last statement would give this method. maybe just inline?